### ISSUE TITLE
Delayed reading of columns

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#45868]
Given this query:

```
SELECT id, view_count, like_count, dislike_count, title FROM youtube ORDER BY view_count DESC LIMIT 100
```

The column `title` is very large, and the query takes a long time:
```
100 rows in set. Elapsed: 26.280 sec. Processed 4.56 billion rows, 500.52 GB (173.38 million rows/s., 19.05 GB/s.)
```

It will be much faster to read the `view_count` column first while keeping track of the data parts and row number in the data part where the records are located, ORDER BY and LIMIT the required number of records, then using the same table snapshot, read the remaining columns.

For example, this query:
```
SELECT view_count FROM youtube ORDER BY view_count DESC LIMIT 100
```

Only takes:
```
100 rows in set. Elapsed: 6.920 sec. Processed 4.56 billion rows, 41.01 GB (658.44 million rows/s., 5.93 GB/s.)
```

**Implementation proposal**

Introduce a new column type, `ColumnLazy`, that will store the reference to the table snapshot to use, the list of data parts, and its main content will be (part num, row num) pairs. Multiple `ColumnLazy` can share the same content (but represent different delayed columns).

If reading is performed from a table of MergeTree family, and the columns are heavily filtered by `ORDER BY ... LIMIT` or in some other way, and the materialization will be possible on the same server, instruct making `ColumnLazy` instead of reading the columns. An additional materialization step will be added to the pipeline.

**Comment by davenger** [comment]
Is it true that if `id` is the PK this effectively will work like the following query?
```
SELECT id, view_count, like_count, dislike_count, title
FROM youtube
WHERE id IN (SELECT id FROM youtube ORDER BY view_count DESC LIMIT 100)
ORDER BY view_count DESC
LIMIT 100
```

**Comment by alexey-milovidov** [comment]
@davenger not quite, because:
- currently, subqueries are using different snapshots to read the table;
- extraneous index analysis will be performed;
- it requires reading the primary key.

**Comment by sundy-li** [comment]
Hits query Q23 can benefit from this.

```
SELECT * FROM hits WHERE URL LIKE '%google%' ORDER BY EventTime LIMIT 10;
```

Since some other DBMS like Apache Doris is using `TopN` optimization. 

Keep EventTime in up to 10 elements min-max heap, if the whole part of `EventTime` did not be capable to change the heap, we can skip reading other columns (including the `URL` column).

**Comment by davenger** [comment]
BTW we already have `_part` and `_part_offset` virtual columns. Sounds like they can be used as (part num, row num) pairs.
Also as `_part` contains full part name it represents table snapshot information.

**Comment by alexey-milovidov** [comment]
@davenger Yes, it can work, but we need to make `_part` `LowCardinality`, and also reference to the table from where the data was read. Maybe a dedicated ColumnLazy will be more natural...

**Comment by zhanglistar** [comment]
What about remote data storage engine like S3 or Hive? Can reading data from remote storage engine get improvement from ColumnLazy?

**Comment by alexey-milovidov** [comment]
@zhanglistar Yes, definitely - the most benefit if the wide format is used (a file per column), and some benefit even if the compact format is used.

**Comment by wudidapaopao** [comment]
I would like to try this.

**Comment by wudidapaopao** [comment]
> Thank you!
> 
> To clarify - this is not an easy task. The overall implementation is imagined as elegant but tricky. There can be a lot of obstacles in this task :)

I will do my best！
